#166383226 Build Identification helper

henryade opened the pull request:
#### What does this PR do?
Adds identification middleware

#### Description of Task to be completed?
Build identification middelware to identify users by id or email

#### How should this be manually tested?
N/A

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#166383226](https://www.pivotaltracker.com/story/show/166383226)

#### Screenshots (if appropriate)
N/A

#### Questions:
N/A

KvNGCzA commented on server/database/models/author.js:
```diff
@@ -1,6 +1,5 @@
 export default (sequelize, DataTypes) => {
   const Author = sequelize.define('Author', {
-    id: DataTypes.INTEGER,
```
Please do not take out the ids.

henryade replied:
> Please do not take out the ids.

Okay. Done

KvNGCzA commented on server/helpers/identification.js:
```diff
@@ -0,0 +1,24 @@
+import { UserRole, User, Role } from '../database/models';
+
+const findIdByEmail = email => User.findOne({
```
This function should take in response as an argument and use it to return proper error messages.

henryade replied:
> This function should take in response as an argument and use it to return proper error messages.

Done

KvNGCzA commented on server/helpers/identification.js:
```diff
@@ -0,0 +1,37 @@
+  where: {
+    email,
+  },
+  attributes: ['id']
```
Why are you returning just the id?

henryade replied:
> Why are you returning just the id?

I am using the email to get the id of the account

KvNGCzA commented on server/helpers/identification.js:
```diff
@@ -0,0 +1,37 @@
+  },
+  attributes: ['id']
+})
+  .then((user) => {
```
use an async function, not then and catch.

KvNGCzA commented on server/helpers/identification.js:
```diff
@@ -0,0 +1,37 @@
+})
+  .then((user) => {
+    if (!user) {
+      return response.status(404).json({
```
use the responseMessage helper function

KvNGCzA commented on server/helpers/identification.js:
```diff
@@ -0,0 +1,37 @@
+import { UserRole, User, Role } from '../database/models';
+
+const findIdByEmail = (email, response) => User.findOne({
```
Rename this function better

KvNGCzA commented on server/helpers/identification.js:
```diff
@@ -0,0 +1,37 @@
+import { UserRole, User, Role } from '../database/models';
```
Have you tested this function?

KvNGCzA commented on server/helpers/identification.js:
```diff
@@ -0,0 +1,37 @@
+  })
+  .catch(error => error);
+
+export default async (param, response) => {
```
rename this function better

KvNGCzA reviewed (changes requested):
Use only destructuring through out the codebase

Adekoreday reviewed (changes requested):
good work

Adekoreday reviewed (changes requested):
in your helper/identification.js
i think you can make your 
include: [
{model: Role, required: true }
]

so it can return user if only its has a role

henryade commented:
> in your helper/identification.js
> i think you can make your
> include: [
> {model: Role, required: true }
> ]
> 
> so it can return user if only its has a role

Done

Ukhu commented on server/helpers/identification.js:
```diff
@@ -0,0 +1,25 @@
+    const { dataValues } = await User.findOne({ where: { email } });
+    return dataValues.id;
+  } catch (error) {
+    responseMessage(response, 404, { message: 'Account not found', error });
```
I don't think you need to return the error object to the user. The message already tells the user what has happened.

henryade replied:
Done

Adekoreday reviewed (changes requested):
nice work

henryade commented:
ok my bad

It's getting it from the userrole table not the role table.

Ukhu commented on server/helpers/responseMessage.js:
```diff
@@ -1,6 +1,6 @@
-export default (response, statusCode, object) => {
+export default (response, statusCode, object, status = 'failure') => {
```
Don't modify this function. You can add the new status to the object you are passing to the function, rather than passing the new status as a separate argument

henryade replied:
If I add it, it will show the failure status first.

Ukhu replied:
Just put it as the first property in the object. It will override the already existing status.

henryade replied:
Done

Ukhu commented on server/helpers/responseMessage.js:
```diff
@@ -1,6 +1,6 @@
-export default (response, statusCode, object) => {
+export default (response, statusCode, object, status = 'failure') => {
   response.status(statusCode).json({
-    status: 'failure',
```
Revert the changes in this function. It shouldn't even show up in your changed files.

henryade replied:
I was asked to use the responseMessage function. The only way to show success is either to create a new function which will lead to duplication or to add that.

Ukhu replied:
Please check my latest comment above

henryade replied:
Done

Ukhu commented on server/helpers/identification.js:
```diff
@@ -0,0 +1,25 @@
+      where: { userId },
+      include: [{ model: Role, required: true }]
+    });
+    return responseMessage(response, 200, { data: dataValues.Role.dataValues }, 'success');
```
Add the status to the object. Don't pass it as a fourth argument.

henryade replied:
If I add it to the object, it will show the failure status first

Ukhu replied:
No, it will override the failure status already there. Just put it as the first property in the object

henryade replied:
Done

JuwonAbiola reviewed (dismissed):
good

KvNGCzA commented on server/helpers/identification.js:
```diff
@@ -0,0 +1,14 @@
+
+const getUserId = async (param, response) => {
+  try {
+    const field = (/\D/g.test(param)) ? { email: param } : { id: param };
```
what happens when there is a number in the email?

henryade replied:
it will still see it as an email

henryade replied:
the only time it regular expression returns false is if the param is a number

KvNGCzA commented on server/helpers/identification.js:
```diff
@@ -0,0 +1,14 @@
+import { User } from '../database/models';
+import responseMessage from './responseMessage';
+
+const getUserId = async (param, response) => {
```
rename this function, and ensure to name the file after its default export

henryade replied:
Done

KvNGCzA commented on server/helpers/identification.js:
```diff
@@ -0,0 +1,14 @@
+const getUserId = async (param, response) => {
+  try {
+    const field = (/\D/g.test(param)) ? { email: param } : { id: param };
+    const user = await User.findOne({ where: field });
```
What happens when this doesn't find a user?

henryade replied:
it returns undefined then whoever wants to use the function will check for undefined.

KvNGCzA commented on server/helpers/identification.js:
```diff
@@ -0,0 +1,14 @@
+import { User } from '../database/models';
```
import `models`, then on a new line destructure `const { User } = models;`

henryade replied:
Done

KvNGCzA commented on server/helpers/index.js:
```diff
@@ -1,7 +1,9 @@
   createToken,
-  responseMessage
+  responseMessage,
+  identify
```
Find a better and more descriptive name than `identify`

henryade replied:
changed

ad3ola committed:
166383226-story(identification): Build identification middleware
build indentification middleware to identify users by email or id
[Starts #166383226]
